### ISSUE TITLE
Nested Navigator does not allow swipe to exit when using `CupertinoPageRoute`

### ISSUE THREAD

**Issue by suesitran** [flutter/flutter#111028]
Github demo: https://github.com/suesitran/flutter_navigation_demo

1. Execute `flutter run` on the code above 
2. In first page, click on 'Go to multitab page'
3. In Page 1, click on 'Go to sub page'
4. In sub-page, swipe from left to right to return to Page 1

**Expected results:** 
App returns to Page 1

**Actual results:**
- in iOS devices, the app returns to the first Single page
- in Android device, unable to swipe to return to previous page

**Comment by exaby73** [comment]
Hello @suesitran. `MaterialPageRoute` by default enables the swipe to go back feature for iOS but not for Android. Maybe what you're looking for is `CupertinoPageRoute`. Closing this issue as this is working as indented. If you disagree, please comment and I can reopen it. Thank you.

**Comment by suesitran** [comment]
hi @exaby73, 

How about the issue on iOS?
* in iOS devices, the app returns to the first Single page

while the expectation is for the App to return to Page 1 - Content 1

**Comment by exaby73** [comment]
I can reproduce this issue on iOS using `MaterialPageRoute` or on any platform using `CupertinoPageRoute`.

### Reproducible on

- `Stable (3.3.0)`
- `Master (3.4.0-19.0.pre.104)`

### Info

<details>
<summary>Video</summary>

<img width="300" src="https://user-images.githubusercontent.com/50555895/188810296-8b346969-cf6f-4a39-bd28-0a91e5c5d5a4.gif">

</details>

<details>
<summary>Code Sample</summary>

```dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Navigation Demo',
      theme: ThemeData(
        primarySwatch: Colors.deepOrange,
      ),
      home: const SinglePage(),
    );
  }
}

class SinglePage extends StatelessWidget {
  const SinglePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) => Scaffold(
        appBar: AppBar(
          title: const Text('Flutter navigation Demo'),
        ),
        body: Center(
          child: ElevatedButton(
            onPressed: () => Navigator.of(context).push(
              CupertinoPageRoute(
                builder: (context) => const MainPage(),
              ),
            ),
            child: const Text('Go to multitab page'),
          ),
        ),
      );
}

class MainPage extends StatefulWidget {
  const MainPage({Key? key}) : super(key: key);

  @override
  State<MainPage> createState() => _MainPageState();
}

class _MainPageState extends State<MainPage> with TickerProviderStateMixin {
  late final TabController _pageController;

  int _currentIndex = 0;

  @override
  void initState() {
    super.initState();

    _pageController = TabController(length: 2, vsync: this)..addListener(() {});
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Navigation demo'),
      ),
      body: TabBarView(
        physics: const NeverScrollableScrollPhysics(),
        controller: _pageController,
        children: const [
          NavigatorMainPage(label: 'Page 1'),
          NavigatorMainPage(label: 'Page 2')
        ],
      ),
      bottomNavigationBar: BottomNavigationBar(
        currentIndex: _currentIndex,
        items: const [
          BottomNavigationBarItem(icon: Icon(Icons.home), label: 'Page 1'),
          BottomNavigationBarItem(icon: Icon(Icons.people), label: 'Page 2')
        ],
        onTap: (value) {
          setState(() {
            _currentIndex = value;
            _pageController.animateTo(_currentIndex);
          });
        },
      ),
    );
  }
}

class NavigatorMainPage extends StatelessWidget {
  const NavigatorMainPage({Key? key, required this.label}) : super(key: key);

  final String label;

  @override
  Widget build(BuildContext context) {
    return Navigator(
      initialRoute: '/',
      onGenerateRoute: _generateRoute,
    );
  }

  Route<dynamic> _generateRoute(RouteSettings settings) {
    if (settings.name == '/') {
      return CupertinoPageRoute(
        builder: (context) => MainPageWithLabel(label: label),
      );
    }

    return CupertinoPageRoute(
      builder: (context) {
        return SubPage(
          label: label,
        );
      },
    );
  }
}

class MainPageWithLabel extends StatelessWidget {
  final String label;

  const MainPageWithLabel({Key? key, required this.label}) : super(key: key);

  @override
  Widget build(BuildContext context) => PageView(
        children: [
          ContentPage(label: '$label - Content 1'),
          ContentPage(label: '$label - Content 2')
        ],
      );
}

class ContentPage extends StatelessWidget {
  final String label;

  const ContentPage({Key? key, required this.label}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          Text(label),
          TextButton(
              onPressed: () => Navigator.of(context).pushNamed('sub page'),
              child: const Text('Go to sub page'))
        ],
      ),
    );
  }
}

class SubPage extends StatelessWidget {
  final String label;

  const SubPage({Key? key, required this.label}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return PageView(
      children: [
        Center(
          child: Text('Sub $label - 1'),
        ),
        Center(
          child: Text('Sub $label - 2'),
        )
      ],
    );
  }
}
```

</details>

### Flutter Doctor

<details><summary><code>flutter doctor -v</code> (Stable)</summary>

```
[✓] Flutter (Channel stable, 3.3.0, on Microsoft Windows [Version 10.0.22000.918], locale en-US)
    • Flutter version 3.3.0 on channel stable at C:\Users\Nabeel\fvm\versions\stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ffccd96b62 (8 days ago), 2022-08-29 17:28:57 -0700
    • Engine revision 5e9e0e0aa8
    • Dart version 2.18.0
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at C:\android-sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = C:\Program Files\BraveSoftware\Brave-Browser\Application\brave.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2022 17.3.2)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.3.32819.101
    • Windows 10 SDK version 10.0.19041.0

[✓] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[✓] VS Code (version 1.71.0)
    • VS Code at C:\Users\Nabeel\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.46.0

[✓] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.22000.918]
    • Chrome (web)      • chrome  • web-javascript • unknown
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 105.0.1343.27

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

<details><summary><code>flutter doctor -v</code> (Master)</summary>

```
[✓] Flutter (Channel master, 3.4.0-19.0.pre.104, on Microsoft Windows [Version 10.0.22000.918], locale en-US)
    • Flutter version 3.4.0-19.0.pre.104 on channel master at C:\Users\Nabeel\fvm\versions\master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 29eb478743 (11 minutes ago), 2022-09-07 01:55:18 -0400
    • Engine revision 2eba5a7485
    • Dart version 2.19.0 (build 2.19.0-171.0.dev)
    • DevTools version 2.17.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at C:\android-sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = C:\Program Files\BraveSoftware\Brave-Browser\Application\brave.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2022 17.3.2)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.3.32819.101
    • Windows 10 SDK version 10.0.19041.0

[✓] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[✓] VS Code (version 1.71.0)
    • VS Code at C:\Users\Nabeel\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.46.0

[✓] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.22000.918]
    • Chrome (web)      • chrome  • web-javascript • unknown
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 105.0.1343.27

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by caseycrogers** [comment]
I'm running into the same issue. The repro above is a bit hard to follow so here's my more minimal repro copied from my dupe-marked issue linked above.

I'd love to see a fix for this-right now we're disabling the iOS back gesture because there are too many places in our app where users would hit this bug if we enabled back swipe.

<details open><summary>Code sample</summary>

```dart
class NestedNavDemo extends StatefulWidget {
  const NestedNavDemo({super.key});

  @override
  State<NestedNavDemo> createState() => _NestedNavDemoState();
}

class _NestedNavDemoState extends State<NestedNavDemo> {
  bool showNestedNav = true;

  @override
  Widget build(BuildContext context) {
    return Navigator(
      onDidRemovePage: (_) => setState(() => showNestedNav = false),
      pages: [
        MaterialPage(
          child: Scaffold(
            body: InkWell(
              onTap: () => setState(() => showNestedNav = true),
              child: const Center(child: Text('Oops, popped to far. Tap to start over.')),
            ),
          ),
        ),
        if (showNestedNav) const MaterialPage(child: InnerNavigator()),
      ],
    );
  }
}

class InnerNavigator extends StatefulWidget {
  const InnerNavigator({super.key});

  @override
  State<InnerNavigator> createState() => _InnerNavigatorState();
}

class _InnerNavigatorState extends State<InnerNavigator> {
  bool showSecondColor = true;

  @override
  Widget build(BuildContext context) {
    const color = Colors.blue;
    const secondColor = Colors.red;

    return Navigator(
      onDidRemovePage: (_) => setState(() => showSecondColor = false),
      pages: [
        CupertinoPage(
          child: AppScaffold(
            appBar: AppBar(),
            body: InkWell(onTap: () => setState(() => showSecondColor = true), child: Container(color: color)),
          ),
        ),
        // Backswiping on this page will dismiss the inner navigator entirely (unexpected).
        if (showSecondColor) CupertinoPage(
          // Tapping the back button in the app bar will only dismiss the second color page (expected).
          child: Scaffold(appBar: AppBar(), body: Container(color: secondColor))
        ),
      ],
    );
  }
}
```

</details>

**Comment by caseycrogers** [comment]
So I dug into this a bunch more. I updated my code sample to include a workaround. It's not too bad. I also root caused the issue. Namely, gestures by default are handled by the lowest widget in the tree first. But, the Cupertino back swipe is intended to take priority over all gesture below it (eg swiping on the edge of the screen should backswipe even if the user drag hit a horizontal scroll view in the screen).
So, it looks like Flutter puts the gesture detector in a stack ABOVE the screen to override the default Flutter behavior:
https://github.com/flutter/flutter/blob/stable/packages/flutter/lib/src/cupertino/route.dart#L834

I'm not totally sure the most elegant way to resolve this conflict is, but we need cupertino gestures to take precedence over their normal gesture detector children, but loose precedence to child cupertino gesture detectors.

<details open><summary>Workaround</summary>

```dart
// Dummy types just to make debugging easier.
class Outer {}

class Inner {}

class NestedNavDemo extends StatefulWidget {
  const NestedNavDemo({super.key});

  @override
  State<NestedNavDemo> createState() => _NestedNavDemoState();
}

class _NestedNavDemoState extends State<NestedNavDemo> {
  bool showNestedNav = true;

  @override
  Widget build(BuildContext context) {
    return Navigator(
      onDidRemovePage: (_) => setState(() => showNestedNav = false),
      pages: [
        CupertinoPage<Outer>(
          child: Scaffold(
            body: InkWell(
              onTap: () => setState(() => showNestedNav = true),
              child: const Center(child: Text('Oops, popped to far. Tap to start over.')),
            ),
          ),
        ),
        if (showNestedNav)
          const CupertinoPage<Outer>(
            child: InnerNavigator(),
            // Uncomment this to disable swiping on the outer page and the inner
            // navigator will behave as expected.
            //fullscreenDialog: true,
          ),
      ],
    );
  }
}

class InnerNavigator extends StatefulWidget {
  const InnerNavigator({super.key});

  @override
  State<InnerNavigator> createState() => _InnerNavigatorState();
}

class _InnerNavigatorState extends State<InnerNavigator> {
  final navigatorKey = GlobalKey<NavigatorState>();
  bool showSecondColor = true;

  @override
  Widget build(BuildContext context) {
    const color = Colors.blue;
    const secondColor = Colors.red;

    final pages = [
      CupertinoPage<Inner>(
        child: Scaffold(
          body: InkWell(onTap: () => setState(() => showSecondColor = true), child: Container(color: color)),
        ),
      ),
      // Backswiping on this page will dismiss the inner navigator entirely (unexpected).
      if (showSecondColor)
        CupertinoPage<Inner>(
          // Tapping the back button in the app bar will only dismiss the second color page (expected).
          child: Scaffold(appBar: AppBar(), body: Container(color: secondColor)),
        ),
    ];

    final canHandlePopInternally = navigatorKey.currentState?.canPop() ?? false;
    return PopScope(
      canPop: !canHandlePopInternally,
      child: Navigator(
        key: navigatorKey,
        onDidRemovePage: (_) => setState(() => showSecondColor = false),
        pages: pages,
      ),
    );
  }
}

```

</details>
